File: packages/python/elizaos/features/advanced_capabilities/providers/relationships.py
from __future__ import annotations

from typing import TYPE_CHECKING

from elizaos.types import Provider, ProviderResult

if TYPE_CHECKING:
    from elizaos.types import IAgentRuntime, Memory, State


def format_relationship(
    relationship: dict[str, str | int | list[str] | dict[str, str]],
    target_name: str,
) -> str:
    tags = relationship.get("tags", [])
    tags_str = (", ".join(tags) if tags else "none") if isinstance(tags, list) else str(tags)

    interactions = relationship.get("metadata", {})
    interaction_count = interactions.get("interactions", 0) if isinstance(interactions, dict) else 0

    return f"- {target_name}: tags=[{tags_str}], interactions={interaction_count}"


async def get_relationships(
    runtime: IAgentRuntime,
    message: Memory,
    state: State | None = None,
) -> ProviderResult:
    entity_id = message.entity_id
    if not entity_id:
        return ProviderResult(
            text="No relationships found.",
            values={"relationshipCount": 0},
            data={"relationships": []},
        )

    try:
        relationships = await runtime.get_relationships({"entityId": str(entity_id)})
    except Exception as e:
        runtime.logger.debug(
            f"Failed to get relationships: src=provider:relationships agentId={runtime.agent_id} error={e}"
        )
        relationships = []

    if not relationships:
        return ProviderResult(
            text="No relationships found.",
            values={"relationshipCount": 0},
            data={"relationships": []},
        )

    def _get_interactions(r: object) -> int:
        if isinstance(r, dict):
            meta = r.get("metadata", {})
            if isinstance(meta, dict):
                val = meta.get("interactions", 0)
                return int(val) if isinstance(val, (int, float)) else 0
        return 0

    sorted_relationships = sorted(
        relationships,
        key=_get_interactions,
        reverse=True,
    )[:30]

    formatted_relationships: list[str] = []
    entity_cache: dict[str, str] = {}
    for rel in sorted_relationships:
        if not isinstance(rel, dict):
            continue
        target_id = rel.get("targetEntityId")
        if not target_id:
            continue

        target_id_str = str(target_id)
        target_name = entity_cache.get(target_id_str)
        if target_name is None:
            target_entity = await runtime.get_entity(target_id_str)
            target_name = target_entity.name if target_entity else target_id_str[:8]
            entity_cache[target_id_str] = target_name

        formatted_relationships.append(format_relationship(rel, target_name))

    if not formatted_relationships:
        return ProviderResult(
            text="No relationships found.",
            values={"relationshipCount": 0},
            data={"relationships": []},
        )

    sender_name = message.content.sender_name if message.content else "Unknown"
    text = f"# {runtime.character.name} has observed {sender_name} interacting with:\n" + "\n".join(
        formatted_relationships
    )

    return ProviderResult(
        text=text,
        values={
            "relationshipCount": len(sorted_relationships),
        },
        data={
            "relationships": sorted_relationships,
        },
    )


relationships_provider = Provider(
    name="RELATIONSHIPS",
    description="Relationships between entities observed by the agent",
    get=get_relationships,
    dynamic=True,
)

File: packages/typescript/src/features/advanced-capabilities/providers/relationships.ts
import { requireProviderSpec } from "../../../generated/spec-helpers.ts";
import type {
	Entity,
	IAgentRuntime,
	Memory,
	Metadata,
	Provider,
	Relationship,
	UUID,
} from "../../../types/index.ts";

// Get text content from centralized specs
const spec = requireProviderSpec("RELATIONSHIPS");

/**
 * Formats the provided relationships based on interaction strength and returns a string.
 * @param {IAgentRuntime} runtime - The runtime object to interact with the agent.
 * @param {Relationship[]} relationships - The relationships to format.
 * @returns {string} The formatted relationships as a string.
 */
/**
 * Asynchronously formats relationships based on their interaction strength.
 *
 * @param {IAgentRuntime} runtime The runtime instance.
 * @param {Relationship[]} relationships The relationships to be formatted.
 * @returns {Promise<string>} A formatted string of the relationships.
 */
async function formatRelationships(
	runtime: IAgentRuntime,
	relationships: Relationship[],
	currentEntityId: UUID,
) {
	// Sort relationships by interaction strength (descending)
	const sortedRelationships = relationships
		.filter((rel) => rel.metadata?.interactions)
		.sort(
			(a, b) =>
				((b.metadata && (b.metadata.interactions as number | undefined)) || 0) -
				((a.metadata && (a.metadata.interactions as number | undefined)) || 0),
		)
		.slice(0, 30); // Get top 30

	if (sortedRelationships.length === 0) {
		return "";
	}

	// Deduplicate target entity IDs to avoid redundant fetches
	const uniqueEntityIds = Array.from(
		new Set(
			sortedRelationships.map(
				(rel) =>
					(rel.sourceEntityId === currentEntityId
						? rel.targetEntityId
						: rel.sourceEntityId) as UUID,
			),
		),
	);

	// Fetch all required entities in a single batch operation
	const entities = await Promise.all(
		uniqueEntityIds.map((id) => runtime.getEntityById(id)),
	);

	// Create a lookup map for efficient access
	const entityMap = new Map<string, Entity | null>();
	entities.forEach((entity, index) => {
		if (entity) {
			entityMap.set(uniqueEntityIds[index], entity);
		}
	});

	const formatMetadata = (metadata?: Metadata) => {
		if (!metadata) return "";
		const lines: string[] = [];
		for (const [key, value] of Object.entries(metadata)) {
			if (value && typeof value === "object") {
				lines.push(`${key}: ${JSON.stringify(value)}`);
			} else {
				lines.push(`${key}: ${String(value)}`);
			}
		}
		return lines.join("\n");
	};

	// Format relationships using the entity map
	const formattedRelationships: string[] = [];
	for (const rel of sortedRelationships) {
		const counterpartEntityId = (
			rel.sourceEntityId === currentEntityId
				? rel.targetEntityId
				: rel.sourceEntityId
		) as UUID;
		const entity = entityMap.get(counterpartEntityId);
		if (!entity) continue;

		const names = entity.names.join(" aka ");
		const tags = rel.tags ? rel.tags.join(", ") : "";
		const metadata = formatMetadata(entity.metadata);
		const parts = [names, tags, metadata].filter((part) => part.length > 0);
		formattedRelationships.push(`${parts.join("\n")}\n`);
	}

	return formattedRelationships.join("\n");
}

/**
 * Provider for fetching relationships data.
 *
 * @type {Provider}
 * @property {string} name - The name of the provider ("RELATIONSHIPS").
 * @property {string} description - Description of the provider.
 * @property {Function} get - Asynchronous function to fetch relationships data.
 * @param {IAgentRuntime} runtime - The agent runtime object.
 * @param {Memory} message - The message object containing entity ID.
 * @returns {Promise<Object>} Object containing relationships data or error message.
 */
const relationshipsProvider: Provider = {
	name: spec.name,
	description: spec.description,
	dynamic: spec.dynamic ?? true,
	get: async (runtime: IAgentRuntime, message: Memory) => {
		// Get all relationships for the current user
		const relationships = await runtime.getRelationships({
			entityIds: [message.entityId],
		});

		if (!relationships || relationships.length === 0) {
			return {
				data: {
					relationships: [],
				},
				values: {
					relationships: "No relationships found.",
				},
				text: "No relationships found.",
			};
		}

		const formattedRelationships = await formatRelationships(
			runtime,
			relationships,
			message.entityId,
		);

		if (!formattedRelationships) {
			return {
				data: {
					relationships: [],
				},
				values: {
					relationships: "No relationships found.",
				},
				text: "No relationships found.",
			};
		}
		return {
			data: {
				relationships: formattedRelationships,
			},
			values: {
				relationships: formattedRelationships,
			},
			text: `# ${runtime.character.name} has observed ${message.content.senderName || message.content.name} interacting with these people:\n${formattedRelationships}`,
		};
	},
};

export { relationshipsProvider };
